### PR TITLE
Make editing the permissions of other users the same as editing your own permissions

### PR DESCRIPTION
**Pull request by chrisroos**
This PR brings the permission editing functionality for other users (under `/users/<id>/`) into line with the functionality we have for editing your own permissions (under `/account/`).

We've copied the relevant controllers and templates from their `/account/` counterparts and modified them to work for editing another user. This has introduced some duplication and also highlighted some minor problems with the original implementation. We'll address those things separately.

## Commits overview

- Show the list of applications that a user does/doesn't have access to
- Allow a user to give another user access to an application
- Allow a user to remove another user's access to an application
- Allow a user to view the permissions a user has for an application
- Allow a user to update the permissions a user has for an application
- Update the "Manage permissions" link on the edit user page so that it points to the new functionality
- Remove the now-redundant functionality for managing user permissions using the table user interface

## Differences from the existing behaviour

The current functionality allows you to grant another user non-signin permissions to an app without also having to grant them the signin permission. The new functionality requires the target user to have the signin permission before any other permissions can be granted. This constraint is currently implemented in the controllers but it would probably be better to move it the models.

## Differences from functionality in the `/account/` namespace

- The routes nested under a specific application (i.e. `/users/<id>/applications/<id>/<path>`) require the user to have the signin permission. This means, for example, that a user's application permissions can only be edited if that user already has access to the application. We'll backport this to the `/account/` namespace separately.
- The signin permission can only be granted for a non-API application. Although not possible through the UI it's technically possible to grant yourself the signin permission to an API application through the equivalent `/account/` functionality.
- The new Pundit policies we've introduced vary from their original counterparts in that they have to take the target user into account. We _should_ be able to replace the account-specific policies with these more generic implementations.

## Routes added

- `GET /users/<id>/applications` - to display two lists of applications: those the user has access to and those the user doesn't have access to.
- `GET /users/<id>/applications/<id>` - we don't have any information to show at this path so we redirect to `/applications` in order to have hackable URLs.
- `POST /users/<id>/applications/<id>/signin_permission` - to add the application signin permission to the user.
- `DELETE /users/<id>/applications/<id>/signin_permission` - to remove the application signin permission from the user.
- `GET /users/<id>/applications/<id>/signin_permission/delete` - to request confirmation that the application signin permission should be removed from the user.
- `GET /users/<id>/applications/<id>/permissions` - to display the list of supported permissions for the application, as well as whether the user has each of those supported permissions.
- `PUT /users/<id>/applications/<id>/permissions` - to modify the users permissions for the application.
- `GET /users/<id>/applications/<id>/permissions/edit` - to display a form allowing the user to choose which of the application's supported permissions the user should have.

These all mirror their `/account/` counterparts.

### Admin managing Org Admin - Before and After

Admin managing permissions for Org Admin - Before

Admin managing permissions for Org Admin - After 1

Admin managing permissions for Org Admin - After 2

Admin managing permissions for Org Admin - After 3

Admin managing permissions for Org Admin - After 4

### Super Org Admin managing Org Admin - Before and After

Super Org Admin managing permissions for Org Admin - Before

Super Org Admin managing permissions for Org Admin - After 1

Super Org Admin managing permissions for Org Admin - After 2

### CODE REVIEW

**Review by floehopper — COMMENTED** [review]
I'm adding this partial review (that I did on Thursday) now in the hope that it might be useful. I will review more of the PR later this morning.

**Review by floehopper — APPROVED** [review]
@chrisroos @chrislo As agreed, given that you both paired on this, I haven't looked at the diff for each commit in detail. However I *have* done the following:
1. Read the PR description carefully
2. Looked through the overall diff fairly carefully
3. Read all the commit notes

In general, it all looks sensible to me - good work! In particular I think the new `UserApplicationPoilcy` is great!

As I discussed on a call with @chrislo, I think it would be worth having a new section or a sentence or two in the PR description explaining the difference between the old behaviour for managing another user's permissions and the new behaviour for managing another user's permissions. If there are no differences, I still think it's important to say that. However, I'm pretty sure there _are_ differences which have essentially been dictated by the new design which doesn't allow a user to give access to an app. I think these are the changes which have forced the changes in the integration tests in [this commit](https://github.com/alphagov/signon/pull/2558/commits/f724bdb98ab1381bb89c26b14fe468cef7373f15). If I've understood things correctly, it's probably also worth explaining that although there are extra limitations on what you can do through the UI, you haven't added any model validations or database constraints which *enforce* these limitations, i.e. you could still add signin permission and other permissions to a user in a Rails console.

Another thing I noticed is that I don't think the new implementation respects the `.permitted_user_params` methods on the sub-classes of `Roles::Base`. I think this used to be enforced in the `UsersController` via the `UserParameterSanitiser`. While it might be a bit academic, because all the sub-classes except `Roles::Normal` include `supported_permission_ids`, I did try to retain this functionality in other similar controllers via `User#permitted_params` e.g. in [`Users::NamesController#user_params`](https://github.com/alphagov/signon/blob/f8494f76a20e1e00abb92a9bb9b5d1c22539d9e8/app/controllers/users/names_controller.rb#L31). At the very least it would be worth mentioning the change in the PR description.

Lastly I wanted to mention that I agree with approach you've taken in relation to `UsersController#require_2sv`. It would be nice to add a comment on [this Trello card](https://trello.com/c/h1CkrB7h) to explain that you've made [some changes](https://github.com/alphagov/signon/pull/2558/commits/df6b4299f5febbe64f5405cf4c854b9638813d9c) in this area.

I'm going to mark this as approved and leave it to your discretion to deal with my feedback!

**Comment by chrisroos** [comment]
Thanks @floehopper! I've rebased and merged and will address your comments separately.

**Comment by chrisroos** [comment]
> Lastly I wanted to mention that I agree with approach you've taken in relation to UsersController#require_2sv. It would be nice to add a comment on [this Trello card](https://trello.com/c/h1CkrB7h) to explain that you've made [some changes](https://github.com/alphagov/signon/pull/2558/commits/df6b4299f5febbe64f5405cf4c854b9638813d9c) in this area.

I've added a [comment to the Trello card](https://trello.com/c/h1CkrB7h/369-prevent-access-to-mandate-2sv-pages-if-organisation-mandates-2sv#comment-65783c7f3fcec3ac67bf491a).

**Comment by chrisroos** [comment]
> As I discussed on a call with @chrislo, I think it would be worth having a new section or a sentence or two in the PR description explaining the difference between the old behaviour for managing another user's permissions and the new behaviour for managing another user's permissions.

I've updated the PR description to explain that users must now have the signin permission before they can be given other non-signin permissions.

File: app/models/user_application_permission.rb [hunk 1]
```diff
@@ -8,6 +8,10 @@ class UserApplicationPermission < ApplicationRecord
 
   before_validation :assign_application_id
 
+  def self.for(user, application)
+    new(user:, application:)
+  end
+
```
[the remark this answers is not in the document]
Comment (chrisroos) [reply app/models/user_application_permission.rb:14]: I agree that `build_for` would probably be better. I've made a note of this and intend to address it as we do more work on this area of the code.

File: test/controllers/users/signin_permissions_controller_test.rb [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+      assert_redirected_to user_applications_path(user)
+    end
+
+    should "prevent unauthenticated users" do
+      user = create(:user)
+      application = create(:application)
+
+      post :create, params: { user_id: user, application_id: application.id }
+
+      assert_redirected_to "/users/sign_in"
```
[the remark this answers is not in the document]
Comment (chrisroos) [reply test/controllers/users/signin_permissions_controller_test.rb:44]: Thanks for the heads up. I'll make this change later.

**Comment by chrisroos** [comment]
> Another thing I noticed is that I don't think the new implementation respects the .permitted_user_params methods on the sub-classes of Roles::Base. I think this used to be enforced in the UsersController via the UserParameterSanitiser. While it might be a bit academic, because all the sub-classes except Roles::Normal include supported_permission_ids, I did try to retain this functionality in other similar controllers via User#permitted_params e.g. in [Users::NamesController#user_params](https://github.com/alphagov/signon/blob/f8494f76a20e1e00abb92a9bb9b5d1c22539d9e8/app/controllers/users/names_controller.rb#L31). At the very least it would be worth mentioning the change in the PR description.

You're correct that we're not using `Roles::<role>.permitted_user_params` in `Users::PermissionsController`. This was previously necessary as permissions were updated, along with other user attributes, in the `UsersController#update` action. Now that permissions are only updated (for other users) in the `Users::PermissionsController` we can rely on the `UserApplicationPermissionsPolicy` to control who's authorised to perform that action.

**Pull request merged**